LifeCycleAssessment_Engine/Compute/EvaluateEnvironmentalProductDeclarationByMass.cs

```diff
@@ -47,44 +49,43 @@ public static partial class Compute
         private static EnvironmentalMetricResult EvaluateEnvironmentalProductDeclarationByMass(IElementM elementM, List<LifeCycleAssessmentPhases> phases, EnvironmentalProductDeclarationField field = EnvironmentalProductDeclarationField.GlobalWarmingPotential, bool exactMatch = false)
         {
             double volume = elementM.ISolidVolume();
-            List<double> epdVal = elementM.GetEvaluationValue(field, phases, QuantityType.Mass, exactMatch);
             List<double> gwpByMaterial = new List<double>();
-            List<double> volumeByRatio = elementM.IMaterialComposition().Ratios.Select(x => volume * x).ToList();
-            List<double> densityOfMassEpd = Query.GetEPDDensity(elementM);
-            List<double> massOfObj = new List<double>();
+            MaterialComposition mc = elementM.IMaterialComposition();
+            List<Material> matList = mc.Materials.ToList();
 
-            if (densityOfMassEpd == null)
+            List<double> epdVals = elementM.GetEvaluationValue(field, phases, QuantityType.Mass, exactMatch);
+            if (epdVals == null || epdVals.Where(x => !double.IsNaN(x)).Sum() <= 0)
             {
-                BH.Engine.Reflection.Compute.RecordError("Density could not be found. Material density is required for all objects using Mass-based evaluations.");
+                BH.Engine.Reflection.Compute.RecordError($"No value for {field} can be found within the supplied EPD.");
                 return null;
             }
 
-            for (int x = 0; x < volumeByRatio.Count; x++)
-            {
-                massOfObj.Add(volumeByRatio[x] * densityOfMassEpd[x]);
-            }
-
-            for (int x = 0; x < epdVal.Count; x++)
+            for (int i = 0; i < matList.Count(); i++)
             {
-                if (double.IsNaN(epdVal[x]))
-                    gwpByMaterial.Add(double.NaN);
-                else
+                double density;
+                List<EnvironmentalProductDeclaration> materialEPDs = matList[i].Properties.OfType<EnvironmentalProductDeclaration>().ToList();
+                if (materialEPDs.Any(x => x.QuantityType == QuantityType.Mass))
                 {
-                    gwpByMaterial.Add(epdVal[x] * massOfObj[x]);
-                }
-            }
+                    double volumeOfMaterial = mc.Ratios[i] * volume;
+                    List<double> densityOfMassEpd = materialEPDs[0].GetEPDDensity();
```

Comment:
Sorry if confusing or not clear, I put this in the wrong place in the diff but referred to it in the comment. 
`List<double> densityOfMassEpd = materialEPDs.Where(x => x.QuantityType == QuantityType.Mass).First().GetEPDDensity()` 

Works the exact same as what's propsed so happy to approve as it. Just wanted to make sure we weren't dropping anything so I made a case with lots of different QTs with good results. Nice one @vgreen-BH!

Reply:
Looking at the [Quantity Type](https://github.com/BHoM/BHoM/blob/master/LifeCycleAssessment_oM/Enums/QuantityType.cs) enum, one option could be `Volume`.

If there are 3 `materialEPDs`, with item 0 being a `volume`, item 1 being a `length`, and item 2 being `mass` then the `if` statement will pass (which is asking if any of the EPD Quantity Types are of type `mass`), but this line will then take the first item which has a quantity type of `volume`.

Therefore, the variable `densityOfMassEpd` is inaccurate, because it would be a `densityOfVolumeEpd` instead.

This is why I think @michaelhoehn suggested code makes more sense - filter down to just the `mass` quantity types, take the first result and then get the density EPD from it. Owing to the way the code is written and how it is read (baring in mind I am not an LCA expert, so I'm just going by what I can see in the code), this will result in more confusion by anyone reading the code later.

Therefore we need to do 2 things in my view:

 - Update the code to use @michaelhoehn suggestion to filter down and take the `mass` type
 - Add code comments that explain what we as developers currently are trying to achieve - that will make it easier in 12 months if we come back to this code to debug to understand what we were thinking at the time we wrote it originally, without risking breaking stuff too much in the future